docs/dev/topology-over-raft.md

```diff
@@ -216,3 +216,94 @@ There are also a few static columns for cluster-global properties:
+the cookie is persisted in the system keyspace and then reloaded on startup
+instead of choosing a random one. The latter is done for the sake
+of idempotency.
+
+*Joining node behavior*
+
+If the node didn't join group 0, it sends `JOIN_NODE_REQUEST` to any node in
+the cluster with the current join cookie and parameters. The receiving node
+can either:
+- Accept the request, meaning that the request was placed in the system.topology
+  and the coordinator will eventually process it,
+- Tell that a request with another join cookie is already in progress for this
+  node and the current node will have to wait until it is processed by the
+  coordinator and removed,
+- Immediately reject the joining node if it is able to verify some parameters
+  on its own and tell that they are incorrect.
+
+In case of other failures (timeouts, connection issues etc.) the request is
+simply retried.
+
+After placing the join request, the node needs to wait until it receives and
+processes a reply. Then, it can start its group 0 server and continue with
+boot.
+
+On receiving `JOIN_NODE_REPLY`, the node first verifies if the cookie included
+in reply matches the cookie that is currently loaded to memory - if it doesn't,
+the response should be rejected with appropriate error variant. Then it should
+process the coordinator's decision:
+- If the new node was rejected, the node should fail the startup procedure.
+- If the new node was accepted:
+  1. The joining node must persist enough information so that it won't change
+     its current parameters in an unsafe way, e.g. features enabled
+     by the cluster must be persisted.
+  2. Features currently enabled by the cluster must be enabled locally.
+  3. The join cookie must be persisted in the system keyspace.
+  4. The node responds with a success.
+
+*Request recipient node behavior*
+
+On receiving `JOIN_NODE_REQUEST`, the node verifies some parameters included
+in request. If the verification fails, a response is immediately sent with
+information about rejection. If the verification is successful, the node
+looks at the `system.topology` under a group 0 guard:
+- If there is already an existing request with the same join cookie, an `ok`
+  response is sent back to the joining node, telling it to wait for a response
+  from the coordinator.
+- If there is already an existing request with a different join cookie, the
+  node is told to retry later after the coordinator notices this request and
+  eventually removes it.
+- If there is no topology request for that node, it is placed in
+  the `system.topology` table and an `ok` response is sent back.
+- If the node has already joined, an `ok` response is sent back.
+
+*Topology coordinator behavior*
+
+Upon noticing a node in `none` state and with `join` topology request, the
+coordinator verifies parameters of the node (e.g. whether it supports all
+of the enabled features) and decides whether to accept it or not. It sends
```

Comment:
I explained the need for initial verification here: https://github.com/scylladb/scylladb/pull/15196#discussion_r1309780775

As for the topology coordinator, we use it to do the "ultimate" verification and marking the node as accepted (i.e. continuing with bootstrapping/replacing the node) because we need to serialize it with marking a feature as enabled, which is currently also done by the topology coordinator.